### PR TITLE
Test when table is created before quota is set

### PR DESCRIPTION
**Pull request by xuebinsu**
This patch adds test cases to ensure that diskquota works when table
is created before quota is set. Specifically,

- When diskquota is preloaded, quota usage is updated automatically
for each active table. Therefore, quotas work once they are set and no
more action is required from the user.
 - When diskquota is NOT preloaded, no active tables will be recorded.
 As a result, when diskquota is preloaded again, the user need to
 SELECT diskquota.init_table_size_table(); and restarts GPDB manually
 in order to update the quota usage.

 This patch also fixes some minor issues in the existing cases.

This is to replace PR #125 .

### CODE REVIEW

File: diskquota.c [hunk 1]
```diff
@@ -1079,3 +1079,49 @@ show_worker_epoch(PG_FUNCTION_ARGS)
...
+		disk_quota_worker_map, (void *) &MyDatabaseId, HASH_FIND, &found);
+	
+	if (found)
+	{
+		need_reload_table_size = workerentry->need_reload_table_size;
+	}
+	LWLockRelease(diskquota_locks.worker_map_lock);
+	if (!found)
+	{
+		ereport(ERROR, (errcode(ERRCODE_INTERNAL_ERROR),
```
Comment (xuebinsu) [inline diskquota.c:1100]: Do NOT throw error here.

File: diskquota.c [hunk 2]
```diff
@@ -1079,3 +1079,49 @@ show_worker_epoch(PG_FUNCTION_ARGS)
...
+		disk_quota_worker_map, (void *) &MyDatabaseId, HASH_FIND, &found);
+	
+	if (found)
+	{
+		workerentry->need_reload_table_size = need_reload_table_size;
+	}
+	LWLockRelease(diskquota_locks.worker_map_lock);
+	if (!found)
+	{
+		ereport(ERROR, (errcode(ERRCODE_INTERNAL_ERROR),
```
Comment (xuebinsu) [inline diskquota.c:1123]: Do NOT throw error here.

File: diskquota.c [hunk 3]
```diff
@@ -1079,3 +1079,49 @@ show_worker_epoch(PG_FUNCTION_ARGS)
...
+						errmsg("[diskquota] worker not found for database \"%s\"",
+							   get_database_name(MyDatabaseId))));	
+	}
+}
```
Comment (xuebinsu) [inline diskquota.c:1127]:
```suggestion
}

```

**Review by beeender — COMMENTED** [review]
I am testing this. Will report back.

File: diskquota.c [hunk 4]
```diff
@@ -1079,3 +1079,49 @@ show_worker_epoch(PG_FUNCTION_ARGS)
 {
 	PG_RETURN_UINT32(worker_get_epoch(MyDatabaseId));
 }
+
+bool
```
[the remark this answers is not in the document]
Comment (xuebinsu) [reply diskquota.c:1083]:
```suggestion
bool
```

File: diskquota.c [hunk 5]
```diff
@@ -1079,3 +1079,49 @@ show_worker_epoch(PG_FUNCTION_ARGS)
...
+bool 
+worker_get_need_reload_table_size()
+{
+	LWLockAcquire(diskquota_locks.worker_map_lock, LW_SHARED);
+
+	bool found = false;
+	bool need_reload_table_size = false;
+	DiskQuotaWorkerEntry * workerentry = (DiskQuotaWorkerEntry *) hash_search(
+		disk_quota_worker_map, (void *) &MyDatabaseId, HASH_FIND, &found);
+
```
[the remark this answers is not in the document]
Comment (xuebinsu) [reply diskquota.c:1092]:
```suggestion

```

File: diskquota.c [hunk 6]
```diff
@@ -1079,3 +1079,49 @@ show_worker_epoch(PG_FUNCTION_ARGS)
...
+	if (found)
+	{
+		need_reload_table_size = workerentry->need_reload_table_size;
+	}
+	LWLockRelease(diskquota_locks.worker_map_lock);
+	if (!found)
+	{
+		ereport(ERROR, (errcode(ERRCODE_INTERNAL_ERROR),
+						errmsg("[diskquota] worker not found for database \"%s\"",
+							   get_database_name(MyDatabaseId))));
```
[the remark this answers is not in the document]
Comment (xuebinsu) [reply diskquota.c:1102]:
```suggestion
							   get_database_name(MyDatabaseId))));
```

File: diskquota.c [hunk 7]
```diff
@@ -1079,3 +1079,49 @@ show_worker_epoch(PG_FUNCTION_ARGS)
...
+	if (found)
+	{
+		workerentry->need_reload_table_size = need_reload_table_size;
+	}
+	LWLockRelease(diskquota_locks.worker_map_lock);
+	if (!found)
+	{
+		ereport(ERROR, (errcode(ERRCODE_INTERNAL_ERROR),
+						errmsg("[diskquota] worker not found for database \"%s\"",
+							   get_database_name(MyDatabaseId))));
```
[the remark this answers is not in the document]
Comment (xuebinsu) [reply diskquota.c:1125]:
```suggestion
							   get_database_name(MyDatabaseId))));
```

File: tests/regress/expected/test_ctas_no_preload_lib.out [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+CREATE ROLE test SUPERUSER;
+SET ROLE test;
+-- Create table with diskquota disabled
+CREATE TABLE t_without_diskquota (i) AS SELECT generate_series(1, 100000)
+DISTRIBUTED BY (i);
+\! gpconfig -c shared_preload_libraries -v 'diskquota'> /dev/null 
+\! gpstop -far > /dev/null
+\c
+-- Make sure that the worker has started
+SELECT diskquota.wait_for_worker_new_epoch();
```
Comment (beeender) [inline tests/regress/expected/test_ctas_no_preload_lib.out:13]: Will be flaky here. No guaranntee registering background worker happened before this.

```
2022-02-15 11:45:12.880071 CST,"cc","contrib_regression",p1267075,th-222197888,"[local]",,2022-02-15 11:45:12 CST,0,con9,cmd2,seg-1,,,,sx1,"ERROR","XX000","[diskquota] worker not found for database ""contrib_regression"" (diskquota.c:1072)",,,,,"PL/pgSQL function diskquota.wait_for_worker_new_epoch() line 6 at assignment","SELECT diskquota.wait_for_worker_new_epoch();",0,,"diskquota.c",1072,"Stack trace:
1    0x55e232767837 postgres errstart + 0x46f
2    0x7f65f5a87428 diskquota.so worker_get_epoch + 0xba
3    0x7f65f5a874b8 diskquota.so show_worker_epoch + 0x1c
4    0x55e23236fcdb postgres <symbol not found> + 0x3236fcdb
5    0x55e232370aea postgres <symbol not found> + 0x32370aea
6    0x7f65d2761d31 plpgsql.so <symbol not found> + 0xd2761d31
7    0x7f65d2761353 plpgsql.so <symbol not found> + 0xd2761353
8    0x7f65d275f627 plpgsql.so <symbol not found> + 0xd275f627
9    0x7f65d275a08c plpgsql.so <symbol not found> + 0xd275a08c
10   0x7f65d2759d20 plpgsql.so <symbol not found> + 0xd2759d20
11   0x7f65d2759b28 plpgsql.so <symbol not found> + 0xd2759b28
12   0x7f65d27598ca plpgsql.so <symbol not found> + 0xd27598ca
13   0x7f65d2757739 plpgsql.so plpgsql_exec_function + 0x397
14   0x7f65d27511a9 plpgsql.so plpgsql_call_handler + 0x1dc
15   0x55e23236fcdb postgres <symbol not found> + 0x3236fcdb
16   0x55e232370aea postgres <symbol not found> + 0x32370aea
17   0x55e23237a095 postgres <symbol not found> + 0x3237a095
18   0x55e23237a741 postgres ExecProject + 0x399
19   0x55e2323aa754 postgres ExecResult + 0x1e8
20   0x55e23236aa19 postgres ExecProcNode + 0x24c
21   0x55e2323646c7 postgres <symbol not found> + 0x323646c7
22   0x55e232360a09 postgres standard_ExecutorRun + 0x465
23   0x55e2323605a1 postgres ExecutorRun + 0x50
24   0x55e2325da7b0 postgres <symbol not found> + 0x325da7b0
25   0x55e2325da3a4 postgres PortalRun + 0x302
26   0x55e2325d1937 postgres <symbol not found> + 0x325d1937
27   0x55e2325d72a4 postgres PostgresMain + 0xd73
28   0x55e232531638 postgres <symbol not found> + 0x32531638
29   0x55e232530b5c postgres <symbol not found> + 0x32530b5c
30   0x55e23252c21c postgres <symbol not found> + 0x3252c21c
"
2022-02-15 11:45:12.880166 CST,"cc","contrib_regression",p1267075,th-222197888,"[local]",,2022-02-15 11:45:12 CST,0,con9,cmd2,seg-1,,,,sx1,"LOG","00000","An exception was encountered during the execution of statement: SELECT diskquota.wait_for_worker_new_epoch();",,,,,,"SELECT diskquota.wait_for_worker_new_epoch();",0,,"elog.c",2115,
2022-02-15 11:45:12.880182 CST,"cc","contrib_regression",p1267075,th-222197888,"[local]",,2022-02-15 11:45:12 CST,0,con9,cmd3,seg-1,,,,sx1,"LOG","00000","statement: SET ROLE test;",,,,,,"SET ROLE test;",0,,"postgres.c",1664,
2022-02-15 11:45:12.887492 CST,"cc","contrib_regression",p1267075,th-222197888,"[local]",,2022-02-15 11:45:12 CST,0,con9,cmd4,seg-1,,,,sx1,"LOG","00000","statement: SELECT diskquota.pause();",,,,,,"SELECT diskquota.pause();",0,,"postgres.c",1664,
2022-02-15 11:45:12.888738 CST,"cc","contrib_regression",p1267075,th-222197888,"[local]",,2022-02-15 11:45:12 CST,0,con9,cmd6,seg-1,,,,sx1,"LOG","00000","statement: SELECT diskquota.wait_for_worker_new_epoch();",,,,,,"SELECT diskquota.wait_for_worker_new_epoch();",0,,"postgres.c",1664,
2022-02-15 11:45:17.012564 CST,,,p1266964,th-222197888,,,,0,con3,cmd1,seg-1,,,,sx1,"WARNING","01000","start workers",,,,,,,0,,"diskquota.c",611,
2022-02-15 11:45:17.012633 CST,,,p1266943,th-222197888,,,,0,,,seg-1,,,,,"LOG","00000","registering background worker ""contrib_regression""",,,,,,,0,,"bgworker.c",347,
2022-02-15 11:45:17.012650 CST,,,p1266943,th-222197888,,,,0,,,seg-1,,,,,"LOG","00000","starting background worker process ""contrib_regression""",,,,,,,0,,"postmaster.c",6167,

```

**Comment by xuebinsu** [comment]
With this PR, `diskquota.state` can now be removed.

**Comment by beeender** [comment]
Xuebin is working on this.

File: tests/regress/expected/test_ctas_no_preload_lib.out (hunk 8 above)
Comment (xuebinsu) [reply tests/regress/expected/test_ctas_no_preload_lib.out:13]: Fixed in PR #144 .

**Review by beeender — APPROVED** [review]

**Review by liuxueyang — APPROVED** [review]

**Review by Sasasu — APPROVED** [review]

**Pull request merged**